server/app/survey.go

```diff
@@ -50,3 +64,197 @@ func (a *UserSurveyApp) StopSurvey(surveyID string) error {
 
 	return nil
 }
+
+func (a *UserSurveyApp) ShouldSendSurvey(userID string, survey *model.Survey) (bool, error) {
+	key := utils.KeyUserSendSurveyLock(userID)
+	locked, err := a.TryLock(key, time.Now().UTC())
```

Comment:
Very insightful! I've refactored this to move the lock acquisition and release process in API layer, however I'm not having it span the duration for checking and sending, rather only the sending function, which should be enough as it performs multiple database calls and creates posts, so should be enough.

The primary reason why I don't want to acquire lock before checking is that if we do so, every time any user loads up MM, we'll create a KV store entry and delete it, for every single MM user on the server, every time they load MM. So, right now, with lock acquisition logic only before the general SendSurvey function, two threads will both see that they both need to send the survey, but only only one will be able to acquire the lock before sending, so the end result is the same minus two DB queries.

Do you feel this might not be sufficient and we still need to expand the lock duration to "check and send"?

Reply:
No, that's still not quite enough because we could still end up with the following series of events that'd result in the user getting two surveys:
1. Goroutine A checks and sees that the survey hasn't been sent
2. Goroutine A acquires the lock
3. Goroutine B checks and sees that the survey hasn't been sent
4. Goroutine A sends the survey and releases the lock
5. Goroutine B acquires the lock, sends the survey, and releases the lock

You don't need to hold the lock to check if the there's a survey open, but you do need to hold it before you check if the survey has already been sent to avoid that. I suppose that you could check if the survey has been sent twice,(once without the lock and then once with it) if you think that trying to acquire the lock will have a performance hit, but I don't expect that to have much of an impact. That also only could have a performance impact if there's a survey open as well, so it's even less likely that there's any effect if people aren't constantly running surveys

Reply:
Makes sense. fixed now! Acquiring lock after check if there is an in progress survey, and releasing it after sending the survey post.